examples/generative/Density estimation using RealNVP.py

```diff
@@ -0,0 +1,216 @@
+    s_layer_3 = keras.layers.Dense(
+        output_dim, activation="relu", kernel_regularizer=regularizers.l2(reg)
+    )(s_layer_2)
+    s_layer_4 = keras.layers.Dense(
+        output_dim, activation="relu", kernel_regularizer=regularizers.l2(reg)
+    )(s_layer_3)
+    s_layer_5 = keras.layers.Dense(
+        2, activation="tanh", kernel_regularizer=regularizers.l2(reg)
+    )(s_layer_4)
+
+    return keras.Model(inputs=input, outputs=[s_layer_5, t_layer_5])
+
+
+Coupling(2).summary()
+"""
+## Real NVP
+"""
+
+
+class RealNVP(keras.Model):
+    def __init__(self, layers_list, masks, distribution, loss_tracker):
+        super(RealNVP, self).__init__()
+
+        self.coupling_layers_num = len(masks)
+
+        # Distribution of the latent space.
+        self.distribution = distribution
+
+        self.masks = masks
+        self.layers_list = layers_list
+        self.loss_tracker = loss_tracker
+
+    def call(self, x, training=True):
+        log_det_inv = 0
+        direction = 1
+        if training:
+            direction = -1
+        for i in range(self.coupling_layers_num)[::direction]:
+            x_masked = x * self.masks[i]
+            reversed_mask = 1 - self.masks[i]
+            s, t = self.layers_list[i](x_masked)
+            s *= reversed_mask
+            t *= reversed_mask
+            gate = (direction - 1) / 2
+            x = (
+                reversed_mask
+                * (x * tf.exp(direction * s) + direction * t * tf.exp(gate * s))
+                + x_masked
+            )
+            log_det_inv += gate * tf.reduce_sum(s, [1])
+
+        return x, log_det_inv
+
+    # Log likelihood of the normal distribution plus the log determinant of the jacobian.
+    def log_loss(self, x):
+        y, logdet = self(x)
+        log_likelihood = self.distribution.log_prob(y) + logdet
+        return -tf.reduce_mean(log_likelihood)
+
+    def train_step(self, data):
+        with tf.GradientTape() as tape:
+            loss = self.log_loss(data)
+
+        g = tape.gradient(loss, self.trainable_variables)
+        self.optimizer.apply_gradients(zip(g, self.trainable_variables))
+        self.loss_tracker.update_state(loss)
+
+        return {"loss": self.loss_tracker.result()}
+
+
+"""
+## Initialization
+"""
+
+# Initializing the masks, the layers and the prior distribution for the RealNVP.
+coupling_layers_num = 6
+
+masks = np.array([[0, 1], [1, 0]] * (coupling_layers_num // 2), dtype="float32")
+
+layers_list = [Coupling(2) for i in range(coupling_layers_num)]
+
+distribution = tfp.distributions.MultivariateNormalDiag(
+    loc=[0.0, 0.0], scale_diag=[1.0, 1.0]
+)
+
+"""
+## Model training
+"""
+loss_tracker = keras.metrics.Mean(name="loss")
+
+optimizer = keras.optimizers.Adam(learning_rate=0.0001)
+
+model = RealNVP(layers_list, masks, distribution, loss_tracker)
+
+model.compile(optimizer=optimizer)
+
+history = model.fit(
+    normalized_data, batch_size=len(normalized_data), epochs=2000, verbose=2
```

Comment:
I also didn't catch this the first time, but you don't do any validation during training. The only model verification you have is a graph at the end. This leads to the question: how did you pick the number of epochs to train for?

Consider monitoring test performance on a subset of the training data. You could do this by implementing `test_step` and passing `validation_split=0.1` to `fit()`

It's also rather weird that the batch size is the data size. What's the motivation for this? Why not use mini batch training?